src/toolkits/drawing_classifier/drawing_classifier.cpp

```diff
@@ -25,20 +27,413 @@ using neural_net::model_backend;
+  }
+
+  // Instantiate the compute context.
+  training_compute_context_ = create_compute_context();
+  if (training_compute_context_ == nullptr) {
+    log_and_throw("No neural network compute context provided");
+  }
+
+  // Report to the user what GPU(s) is being used.
+  std::vector<std::string> gpu_names = training_compute_context_->gpu_names();
+  print_training_device(gpu_names);
+
+  // Set additional model fields.
+  add_or_update_state({
+      {"num_classes", training_data_iterator_->class_labels().size()},
+      {"training_iterations", 0},
+  });
+
+  // Initialize the neural net. Note that this depends on statistics computed by
+  // the data iterator.
+  bool use_random_init = true;
+  nn_spec_ = init_model(use_random_init);
+
+  // TODO: Do not hardcode values
+  training_model_ = training_compute_context_->create_drawing_classifier(
+      /* TODO: nn_spec_->export_params_view().
+       * Until the nn_spec in C++ isn't ready, do not pass in any weights.
+       */
+      read_state<size_t>("batch_size"), read_state<size_t>("num_classes"));
+
+  // Print the header last, after any logging triggered by initialization above.
+  if (training_table_printer_) {
+    training_table_printer_->print_header();
+  }
+}
+
+
+// Returns the validation accuracy and validation loss respectively as a tuple
+std::tuple<float, float> drawing_classifier::compute_validation_metrics(
+    size_t num_classes, size_t batch_size) {
+
+  float cumulative_val_loss = 0.f;
+  size_t val_size = 0;
+  size_t val_num_correct = 0;
+  size_t val_num_samples = 0;
+  validation_data_iterator_->reset();
+
+  // To support double buffering, use a queue of pending inference results.
+  std::queue<result> pending_batches;
+
+  auto pop_until_size = [&](size_t remaining) {
+
+    while (pending_batches.size() > remaining) {
+
+      // Pop one batch from the queue.
+      result batch = pending_batches.front();
+      pending_batches.pop();
+
+      size_t batch_num_correct = 0;
+      batch_num_correct = static_cast<size_t>(
+        *(batch.accuracy_info.data()) * batch.data_info.num_samples);
+      val_num_correct += batch_num_correct;
+      val_num_samples += batch.data_info.num_samples;
+      float val_loss = std::accumulate(batch.loss_info.data(),
+                            batch.loss_info.data() + batch.loss_info.size(),
+                            0.f, std::plus<float>());
+      cumulative_val_loss += val_loss;
+    }
+  };
+
+  while (validation_data_iterator_->has_next_batch()) {
+
+    // Wait until we have just one asynchronous batch outstanding. The work
+    // below should be concurrent with the neural net inference for that batch.
+    pop_until_size(1);
+
+    result result_batch;
+    result_batch.data_info = validation_data_iterator_->next_batch(batch_size);
+
+    // Submit the batch to the neural net model.
+    std::map<std::string, shared_float_array> results =
+        training_model_->predict({
```

Comment:
is the predict not plugged in your test mock?

Reply:
Good catch, thanks for pointing out!

Reply:
I didn't add it because we don't expect the validation data to be non-empty. However, I agree that it's still good to mock predict though. Will add that.

Reply:
well, let's have a common until file to aggregate all data-generators for our mock purposes.

Reply:
After discussing offline, let's bring this in a later PR.